url2698a2c731_cassandra_project_cassandra_stand_alone_tgz-pJ8-eu_cassandra_sim_MySimpleDynamicScenarioSimulationJ8/src/eu/cassandra/sim/utilities/DerbyResults.java

```diff
@@ -0,0 +1,565 @@
+	 */
+	@Override
+	public void addKPIs(String inst_id, double maxPower, double avgPower, double energy, double cost, double co2) {
+		boolean first = false;
+		
+		String tableName;
+		String tick_tableName;
+		String id;
+		
+		if(inst_id.equalsIgnoreCase(AGGR)) {
+			id = AGGR;
+			tableName = COL_AGGRKPIS;
+			tick_tableName = COL_AGGRRESULTS;
+		} else {
+			id = inst_id;
+			tableName = COL_INSTKPIS;
+			tick_tableName = COL_INSTRESULTS;
+		}
+				
+		double[] data = getKPIData(tableName, "inst_id", id);
+		
+		double newMaxPower = maxPower;
+		double newAvgPower = avgPower;
+		double newEnergy = energy;
+		double newCost = cost;
+		double newCo2 = co2;
+		if(data == null) 
+			first = true;
+		else {
+			newMaxPower += data[0];
+			newAvgPower += data[1];
+			newEnergy += data[2];
+			newCost +=data[3];
+			newCo2 += data[4];
+		}
+		
+		double maxavgValue = newAvgPower;
+		
+		String maxQuery = "SELECT MAX(P) AS maxp FROM " + tick_tableName;
+		if(!inst_id.equalsIgnoreCase(AGGR))
+			maxQuery += " WHERE inst_id='" + inst_id +"'";
+		ResultSet maxavg = executeSelectQuery(maxQuery);
+		if(maxavg != null)
+			try {
+				while (maxavg.next())
+					maxavgValue = maxavg.getDouble("maxp");
+			} catch (SQLException e) {
+				System.out.println(" . . . exception thrown:");
+				errorPrint(e);
+			}
+		
+		String psQuery = "";
+		if(first) 
+			psQuery = "insert into " + tableName + " values('" + id + "'," + maxavgValue + ", " + newMaxPower + ", " + newAvgPower + ", " + newEnergy + ", " + newCost + ", " + newCo2+ ")";
+		else 
+			psQuery = "update " + tableName + " set AVGPEAK=" + maxavgValue +", MAXPOWER=" + newMaxPower + ", AVGPOWER=" + newAvgPower 
+					+ ", ENERGY=" + newEnergy + ", COST=" + newCost + ", CO2=" + newCo2 + " where inst_id='" + id + "'";
+		executeUpdateQuery(psQuery);
+	
+	}
+
+	/* (non-Javadoc)
+	 * @see eu.cassandra.sim.utilities.DBResults#addAppKPIs(java.lang.String, double, double, double, double, double)
+	 */
+	@Override
+	public void addAppKPIs(String app_id, double maxPower, double avgPower, double energy, double cost, double co2) {
+		String psQuery = "";
+		String tableName = COL_APPKPIS;
+		double[] data = getKPIData(tableName, "app_id", app_id);
+		
+		double newMaxPower = maxPower;
+		double newAvgPower = avgPower;
+		double newEnergy = energy;
+		double newCost = cost;
+		double newCo2 = co2;
+		if(data == null) {
+			psQuery = "insert into " + tableName + " values('" + app_id + "'," + newMaxPower + ", " + newAvgPower + ", " + newEnergy + ", " + newCost + ", " + newCo2+ ")";
+			
+		} else {
+			newMaxPower += data[0];
+			newAvgPower += data[1];
+			newEnergy += data[2];
+			newCost += data[3];
+			newCo2 += data[4];
+			psQuery = "update " + tableName + " set MAXPOWER=" + newMaxPower + ", AVGPOWER=" + newAvgPower 
+					+ ", ENERGY=" + newEnergy + ", COST=" + newCost + ", CO2=" + newCo2 + " where app_id='" + app_id + "'";
+		}		
+		executeUpdateQuery(psQuery);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 143:164, 172:193.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.